tests/contrib/autoname/test_glom_named.py

```diff
@@ -0,0 +1,49 @@
+import pytest
+
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+
+from pyro.contrib.autoname import glom_name
+
+
+class B(object):
+    def __init__(self):
+        b = 3
+
+
+def test_dynamic_simple():
+
+    @glom_name
+    def model():
+
+        x = pyro.sample(dist.Bernoulli(0.5))
+        y = pyro.sample(dist.Bernoulli(0.5))
+        z = [x, y, None]
+        i = 2
+        z[i] = pyro.sample(dist.Bernoulli(0.5))
+        zz = {}
+        zz["a"] = pyro.sample(dist.Bernoulli(0.5))
+        ab = B()
+        ab.b = pyro.sample(dist.Bernoulli(0.5))
+        zz["c"] = [B(), None, None, None]
+        zz["c"][0].b = pyro.sample(dist.Bernoulli(0.5))
+        for j in range(1, 3):
+            zz["c"][j] = pyro.sample(dist.Bernoulli(0.5))
```

Comment:
This is really nice. How smart is it? Can we
```py
for key, value in my_dict.items():
    zz["{}_{}".format(self.prefix, key)] = pyro.sample(dist.Normal(value, 1.))
```

Reply:
Yes, that would work exactly as expected.  The main limitation of this autonaming scheme is that, as currently written, it doesn't support closures due to the way the recompilation works.  For example, note that in this test, the class `B` has to be defined at the module level rather than inside the test function.  I'm guessing there's a workaround for this shortcoming, though.